src/lib/eap/base.c

```diff
@@ -386,80 +388,81 @@ void eap_add_reply(request_t *request, fr_dict_attr_t const *da, uint8_t const *
 	REXDENT();
 }
 
-/** Run a subrequest through a virtual server, managing the eap_session_t of the child
+/** Handle the result of running a subrequest through a virtual server
  *
- * If eap_session_t has a child, inject that into the request.
- *
- * If after the request has run, the child eap_session_t is no longer present,
- * we assume it has been freed, and fixup the parent eap_session_t.
+ * Storing the value of the State attribute in readiness for the next round.
+ */
+static unlang_action_t eap_virtual_server_resume(UNUSED rlm_rcode_t *p_result, UNUSED int *priority,
+						 request_t *request, void *uctx)
+{
+	eap_session_t	*eap_session = talloc_get_type_abort(uctx, eap_session_t);
+	fr_pair_t	*vp;
+
+	/*
+	 *	Child request stores it's own session state - including the eap_session_t
+	 *	Keep a copy of the session ID, then it can be presented back to the
+	 *	child on the next round.
+	 */
+	TALLOC_FREE(eap_session->child_state);
+	vp = fr_pair_find_by_da(&request->reply_pairs, NULL, attr_state);
+	if (vp) {
+		RDEBUG4("Storing child state %pV", &vp->data);
+		eap_session->child_state = talloc_typed_memdup(eap_session, vp->vp_octets, vp->vp_length);
+	}
+
+	return UNLANG_ACTION_CALCULATE_RESULT;
+}
+
+/** Run a subrequest through a virtual server
  *
- * If the eap_session_t pointer changes, this is considered a fatal error.
+ * If eap_session_t has a child_state, inject that as an attribute in the request.
  *
  * @param[in] request		the current (real) request.
  * @param[in] eap_session	representing the outer eap method.
  * @param[in] virtual_server	The default virtual server to send the request to.
- * @return the rcode of the last executed section in the virtual server.
+ * @return
+ * 	- UNLANG_ACTION_PUSHED_CHILD on success
+ *	- UNLANG_ACTION_FAIL on error
  */
-rlm_rcode_t eap_virtual_server(UNUSED request_t *request, UNUSED eap_session_t *eap_session, UNUSED char const *virtual_server)
+unlang_action_t eap_virtual_server(request_t *request, eap_session_t *eap_session, char const *virtual_server)
 {
-#if 1
-	return RLM_MODULE_FAIL;
-#else
-	eap_session_t	*eap_session_inner;
-	rlm_rcode_t	rcode;
-	fr_pair_t	*vp;
-	CONF_SECTION	*server_cs;
+	virtual_server_t const	*vs;
+	CONF_SECTION		*server_cs;
+	fr_pair_t		*vp;
 
-	vp = fr_pair_find_by_da(&request->control_pairs, NULL, attr_virtual_server);
-	server_cs = vp ? virtual_server_find(vp->vp_strvalue) : virtual_server_find(virtual_server);
+	fr_assert(request->parent);
 
-	if (server_cs) {
-		RDEBUG2("Running request through virtual server \"%s\"", cf_section_name2(unlang_call_current(request)));
-	} else {
-		RDEBUG2("Running request in virtual server");
-	}
+	vs = virtual_server_find(virtual_server);
```

Comment:
Might be something to fix.  I think in a lot of places we resolve virtual servers on startup now.  There's actually a struct that represents a virtual server, whereas there didn't use to be.